### ISSUE TITLE
APICTL 4.0 - error while importing API when  the max revision numbers is reached

### ISSUE THREAD

**Issue by JolantaZ** [wso2/api-manager#1286]
### Description

Issue - There is an error while importing/deploying the APIs using the APICTL tool in the APIM-4.0.0 with MSSql DB.
Error occure when the number of revisions have reached a maximum of 5.

Error in terminal:
Error importing API.
Status: 500
Response: {"code":900967,"message":"General Error","description":"Server Error Occurred","moreInfo":"","error":[]}
apictl: Error importing API Reason: 500
Exit status 1

Logs from Api Manager points to problem with sql syntax and usage LIMIT operator which is not supported by SQL Server.

### Steps to Reproduce

1. Configure APIM-4.0.0 for Docker with MsSql DB
2. Pick the API with 5 deployment revisions reached.
3. Import the API using the APICTL tool >> 
commands:
apictl import api with --rotate-revision parameter
or
apictl vcs deploy ( parameter rotateRevision: true in api_meta.yaml file) 
4. Error occure

### Affected Component

APICTL

### Version

apictl 4.0.1, Api Manager 4.0

### Environment Details (with versions)

Linux, Ubuntu 20.04.5, MS SqlServer Database

### Relevant Log Output

```shell
[2023-01-18 15:58:27,537]  INFO - ImportUtils Existing API found, attempting to update it...
[2023-01-18 15:58:31,341] ERROR - ApiMgtDAO Failed to get the earliest revision for api ID: 0874d84e-533d-4a54-acc2-014567c9c8cf
com.microsoft.sqlserver.jdbc.SQLServerException: Incorrect syntax near 'LIMIT'.
        at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:265) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1673) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:620) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:540) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7627) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3912) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:268) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:242) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeQuery(SQLServerPreparedStatement.java:459) ~[mssql-jdbc-11.2.0.jre11.jar:?]
        at jdk.internal.reflect.GeneratedMethodAccessor62.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:114) ~[jdbc-pool_9.0.35.wso2v1.jar:?]
        at com.sun.proxy.$Proxy53.executeQuery(Unknown Source) ~[?:?]
        at org.wso2.carbon.apimgt.impl.dao.ApiMgtDAO.getEarliestRevision_aroundBody768(ApiMgtDAO.java:16079) [org.wso2.carbon.apimgt.impl_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.impl.dao.ApiMgtDAO.getEarliestRevision(ApiMgtDAO.java:16072) [org.wso2.carbon.apimgt.impl_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.impl.APIProviderImpl.getEarliestRevisionUUID_aroundBody580(APIProviderImpl.java:9251) [org.wso2.carbon.apimgt.impl_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.impl.APIProviderImpl.getEarliestRevisionUUID(APIProviderImpl.java:9250) [org.wso2.carbon.apimgt.impl_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.impl.UserAwareAPIProvider.getEarliestRevisionUUID(UserAwareAPIProvider.java:1) [org.wso2.carbon.apimgt.impl_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.rest.api.publisher.v1.common.mappings.ImportUtils.importApi(ImportUtils.java:335) [org.wso2.carbon.apimgt.rest.api.publisher.v1.common_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.rest.api.publisher.v1.common.ImportExportAPIServiceImpl.importAPI(ImportExportAPIServiceImpl.java:191) [org.wso2.carbon.apimgt.rest.api.publisher.v1.common_9.0.174.jar:?]
        at org.wso2.carbon.apimgt.rest.api.publisher.v1.impl.ApisApiServiceImpl.importAPI(ApisApiServiceImpl.java:3905) [classes/:?]
        at org.wso2.carbon.apimgt.rest.api.publisher.v1.ApisApi.importAPI(ApisApi.java:1192) [classes/:?]
        at jdk.internal.reflect.GeneratedMethodAccessor1498.invoke(Unknown Source) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179) [cxf-core-3.3.7.jar:3.3.7]
        at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96) [cxf-core-3.3.7.jar:3.3.7]
        at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201) [cxf-rt-frontend-jaxrs-3.3.7.jar:3.3.7]
        at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104) [cxf-rt-frontend-jaxrs-3.3.7.jar:3.3.7]
        at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59) [cxf-core-3.3.7.jar:3.3.7]
        at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96) [cxf-core-3.3.7.jar:3.3.7]
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308) [cxf-core-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121) [cxf-core-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:296) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:215) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:660) [tomcat-servlet-api_9.0.34.wso2v1.jar:?]
        at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:271) [cxf-rt-transports-http-3.3.7.jar:3.3.7]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [tomcat_9.0.34.wso2v1.jar:?]
        at org.wso2.carbon.identity.context.rewrite.valve.TenantContextRewriteValve.invoke(TenantContextRewriteValve.java:107) [org.wso2.carbon.identity.context.rewrite.valve_1.4.25.jar:?]
        at org.wso2.carbon.identity.authz.valve.AuthorizationValve.invoke(AuthorizationValve.java:110) [org.wso2.carbon.identity.authz.valve_1.4.25.jar:?]
        at org.wso2.carbon.identity.auth.valve.AuthenticationValve.invoke(AuthenticationValve.java:102) [org.wso2.carbon.identity.auth.valve_1.4.25.jar:?]
        at org.wso2.carbon.tomcat.ext.valves.CompositeValve.continueInvocation(CompositeValve.java:101) [org.wso2.carbon.tomcat.ext_4.6.2.jar:?]
        at org.wso2.carbon.tomcat.ext.valves.TomcatValveContainer.invokeValves(TomcatValveContainer.java:49) [org.wso2.carbon.tomcat.ext_4.6.2.jar:?]
        at org.wso2.carbon.tomcat.ext.valves.CompositeValve.invoke(CompositeValve.java:62) [org.wso2.carbon.tomcat.ext_4.6.2.jar:?]
        at org.wso2.carbon.tomcat.ext.valves.CarbonStuckThreadDetectionValve.invoke(CarbonStuckThreadDetectionValve.java:145) [org.wso2.carbon.tomcat.ext_4.6.2.jar:?]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690) [tomcat_9.0.34.wso2v1.jar:?]
        at org.wso2.carbon.tomcat.ext.valves.CarbonContextCreatorValve.invoke(CarbonContextCreatorValve.java:57) [org.wso2.carbon.tomcat.ext_4.6.2.jar:?]
        at org.wso2.carbon.tomcat.ext.valves.RequestCorrelationIdValve.invoke(RequestCorrelationIdValve.java:126) [org.wso2.carbon.tomcat.ext_4.6.2.jar:?]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1590) [tomcat_9.0.34.wso2v1.jar:?]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat_9.0.34.wso2v1.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat_9.0.34.wso2v1.jar:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
[2023-01-18 15:58:31,385] ERROR - GlobalThrowableMapper Error while importing API: SelfCareAPI-Test version: v1.0 Failed to get the earliest revision for api ID: 0874d84e-533d-4a54-acc2-014567c9c8cf
```

### Related Issues

Very similar issue with operator LIMIT reported here https://github.com/wso2/api-manager/issues/1021

### Suggested Labels

Priority/High, Severity/Critical

**Comment by lakshithagunasekara** [comment]
https://github.com/wso2/api-manager/issues/1021 and https://github.com/wso2/api-manager/issues/1286 are similar and @IsuruMaduranga will work on both issues.

**Comment by chamilaadhi** [comment]
Tried api with 5 revisions and no issue found in the latest 4.2.0
![Screenshot 2023-03-01 at 10 58 28 AM](https://user-images.githubusercontent.com/4861150/222053923-9c244983-8e91-4afc-a5fa-67a5c5c4e267.png)

This is fixed from https://github.com/wso2/carbon-apimgt/pull/11176

**Comment by JolantaZ** [comment]
@chamilaadhi is it fixed in this release? https://github.com/wso2/product-apim/releases/tag/v4.0.0
Is it possible to apply the fix to this version 4.0.0 by just copy some files to my already created and running Api Manager?

**Comment by chamilaadhi** [comment]
@JolantaZ 

We do not modify released tags. So this fix is not available in the 4.0.0 release tag https://github.com/wso2/product-apim/releases/tag/v4.0.0. If you need to have this fix in 4.0.0, you will have to update the product (see https://wso2.com/updates/). If not, to manually apply this fix to the pack, first you need to apply the https://github.com/wso2/carbon-apimgt/pull/11176 fix to https://github.com/wso2/carbon-apimgt/releases/tag/v9.0.174 tag first and build that repo and then build the https://github.com/wso2/product-apim/releases/tag/v4.0.0 tag to get a 4.0.0 pack with this fix. I would recommend updating the product instead of building it by applying the fix manually since you would get fixes for other issues.

This fix is already added to the 4.1.0 public release (https://github.com/wso2/product-apim/releases/tag/v4.1.0). You could migrate your 4.0.0 version to 4.1.0 to get this as well.
